src/algorithms/synthesis/dd_synthesis.cpp

```diff
@@ -85,6 +85,39 @@ namespace syrec {
         cube.pop_back();
     }
 
+    //please refer to the second optimization approach introduced in https://www.cda.cit.tum.de/files/eda/2017_rc_improving_qmdd_synthesis_of_reversible_circuits.pdf
+    auto DDSynthesizer::finalSrcPathSignature(dd::mEdge const& src, dd::mEdge const& current, TruthTable::Cube::Vector const& p1SigVec, TruthTable::Cube::Vector const& p2SigVec, std::unique_ptr<dd::Package<>>& dd) const -> TruthTable::Cube::Vector {
+        if (current.p->v == src.p->v || current.p->v == 0U) {
+            TruthTable::Cube::Vector rootSigVec;
+            pathFromSrcDst(src, current.p, rootSigVec);
+            return rootSigVec;
+        }
+
+        TruthTable::Cube::Vector rootSigVec;
+        pathFromSrcDst(src, current.p, rootSigVec);
+
+        const auto tables = dd->mUniqueTable.getTables();
+
+        auto const& table = tables[current.p->v];
+
+        for (auto* p: table) {
+            if (p != nullptr && p != current.p) {
```

Comment:
In order to not waste computational effort you should also check whether the node has a non-zero ref count here.
The unique table might still contain nodes from previous computations that have not yet been garbage collected.
A simple `&& p->ref > 0` should do.